P4-OVS: Introduce support for tap port creation

nupuruttarwar opened the pull request:
This patch introduces support required for tap port creation and
GNMI cli enhancements for pipeline name, mtu value and mempool
name DPDK port configuration.

Example commands:
With all parameters: sudo gnmi-cli set "device:virtual-device,\
name:TAP0,pipeline-name:pipe,mempool-name:MEMPOOL0,mtu:1500,port-type:TAP"
Default parameters skipped: sudo gnmi-cli set "device:virtual-device,\
name:TAP1,mtu:1500,port-type:TAP"

Title: Introduce support for tap port creation
Change-type: FeatureImplementation
Signed-off-by: nupuruttarwar <nupur.uttarwar@intel.com>

nupuruttarwar committed:
P4-OVS: Introduce support for tap port creation

This patch introduces support required for tap port creation and
GNMI cli enhancements for pipeline name, mtu value and mempool
name DPDK port configuration.

Example commands:
With all parameters: sudo gnmi-cli set "device:virtual-device,\
name:TAP0,pipeline-name:pipe,mempool-name:MEMPOOL0,mtu:1500,port-type:TAP"
Default parameters skipped: sudo gnmi-cli set "device:virtual-device,\
name:TAP1,mtu:1500,port-type:TAP"

Title: Introduce support for tap port creation
Change-type: FeatureImplementation
Signed-off-by: nupuruttarwar <nupur.uttarwar@intel.com>

stolsma commented:
Nice! Will have a look after the weekend but don't wait for me if you want to merge!👍

stolsma commented on external/PATCH-01-STRATUM:
```diff
@@ -57,7 +57,7 @@ index e1f4338e..bffdc71e 100644
 +                {
 +                    "program-name": "dummy",
-+		    "cpu_numa_node": "0",
++                    "cpu_numa_node": "0",
```
Could all the formatting changes not related to implementing TAP support be included in another PR? It is very difficult to see which code changed and to get to understand relations between code when there is so much clutter....

nupuruttarwar replied:
@stolsma thanks for the review!! Since it was a minor change, I added it here. But you are right, I can create a separate PR for such changes from next time

stolsma commented on external/PATCH-01-STRATUM:
```diff
@@ -692,22 +707,23 @@ index f553054d..3f41e9e1 100644
-+  PORT_TYPE_LINK = 2;
-+  PORT_TYPE_SOURCE = 3;
-+  PORT_TYPE_SINK = 4;
++  PORT_TYPE_VHOST = 1;
```
Does the change from LINK to VHOST naming also means the code and documentation in IPDK need to be changed?

nupuruttarwar replied:
That's not required. Internal logic introduced on line 284 in gnmi_cli.cc file will change the port type to vhost if device-type is virtual device and port type is link. This is to avoid confusion for existing users. Also, both vhost user ports and pci link ports are treated as link ports by DPDK backend. I am working on documentation for tap ports, will update the PR soon

stolsma commented on p4proto/bfIntf/bf_chassis_manager.cc:
```diff
@@ -158,17 +176,68 @@ ::util::Status BfChassisManager::ValidateAndAdd(uint64 node_id, uint32 port_id,
-      RETURN_IF_ERROR(AddPortHelper(node_id, unit, sdk_port_id, singleton_port, &config));
-  }
+    if ((validate & GNMI_CONFIG_PORT_TYPE) == GNMI_CONFIG_PORT_TYPE) {
+      if (((config.port_type == PORT_TYPE_VHOST) &&
```
Also LINK to VHOST question related to operational use...

nupuruttarwar replied:
Sorry, I didn't understand the question here?

stolsma commented on p4proto/bfIntf/bf_chassis_manager.h:
```diff
@@ -25,10 +25,29 @@
+// Independant TAP ports shouldn't have the below params.
+#define GNMI_CONFIG_UNSUPPORTED_MASK_TAP (GNMI_CONFIG_DEVICE_TYPE | GNMI_CONFIG_QUEUE_COUNT | \
+                                          GNMI_CONFIG_SOCKET_PATH | GNMI_CONFIG_HOST_NAME)
+
```
Why use a negative mask instead of defining a seperate GNMI_CONFIG_TAP mask defining the TAP parameters? That will make it easier to add other link types later on with other specific parameters if needed ?

stolsma commented on utilities/gnmi_cli.cc:
```diff
@@ -171,6 +171,15 @@ bool extract_interface_node(char **path, char *node_path) {
+          //Validate the device value
+          if((strcmp(value,"virtual-device") != 0) &&
+             (strcmp(value,"physical-device") != 0)) {
+              return -1;
```
Will this `virtual-device` and  `physical-device` option be documented somewhere? Or is it already documented and I don't know it exists?

nupuruttarwar replied:
Will remove physical device since this is for pci based ports, support for which is being still worked upon. Once this support is available, will surely add the documentation :)

stolsma reviewed (changes requested):
Looks good in general, I have mostly some questions about changing cli use and documenting it. Can't wait until it is merged! 😃

nupuruttarwar committed:
P4-OVS: Update documentation for TAP port creation

This patch updates documentation for TAP port creation and
addresses comments

Title: Update documentation for TAP port creation
Change-type: Other
Signed-off-by: nupuruttarwar <nupur.uttarwar@intel.com>
